text/0000-better-env-handling.md

```diff
@@ -0,0 +1,89 @@
+- Start Date: 2020-04-23
+- Relevant Team(s): Ember.js, Learning
+- RFC PR: [#619](https://github.com/emberjs/rfcs/pull/619)
+- Tracking: (leave this empty)
+
+# Better environment handling
+
+## Summary
+
+This RFC proposes to add a tiny module similar to `@glimmer/env`, to allow to more easily work with common environment questions.
+
+## Motivation
+
+This RFC tries to solve two common problems:
+
+1 - **Handling test-specific code**.
+
+The old way of doing this, `Ember.testing`, is not really ideal and recommended anymore.
+
+Example use cases where you might want to use this:
+
+- Disabling polling code in tests
+- Preventing test-breaking code like `window.reload()`
+- Preventing log output in tests
+
+2 - **Handling debug-specific (=non-production) code**.
+
+You can use `import { DEBUG } from "@glimmer/env"` for this, or `import { runInDebug } from "@ember/debug"`, although the later [seems to be buggy as of now](https://github.com/emberjs/ember.js/issues/18912).
+
+An exmaple use case where you might want to use this:
+
+- Auto-signing in a mocked user in development environment
+
+This RFC proposes to provide a more more streamlined, approved and documented way to handle these cases, by introducing an `@ember/env` namespace which can be used like this:
+
+```js
+import { DEBUG, TESTING } from "@ember/env";
+```
+
+## Detailed design
+
+```js
+import { DEBUG } from "@ember/env";
+```
+
+This works exactly as `import { DEBUG } from "@glimmer/env"`.
+The reasoning for moving this into the `@ember` namespace is that it is not clear why this should be imported from glimmer, which is the rendering engine.
+
+```js
+import { TESTING } from "@ember/env";
+```
+
+This works like `Ember.testing`.
```

Comment:
I actually think this is a bit off from a design perspective:

* `Ember.testing` **cannot** be checked in module scope, it is only set when starting an individual test and unset when the test is completed.
* We want this value to mean something more akin to "this is a testing enabled build"

In other words, we do **not** want this to be the same as `Ember.testing` we want it to be:

* constant (does not change / live update without a refresh)
* `true` for all of the following:
  * `ember test`
  * `ember test --environment=production`
  * `ember serve` + `/tests/`
* `false` for all of the following:
  * `ember serve` + any non `/tests/` path
  * `ember build --environment=production` (non-testing related production build)

Reply:
What is the difference between this and `environment === 'test'` ?

Reply:
Not much from an applications perspective, but addons cannot use `import config from '../config/environment'` (because they have no clue what the app name they are consumed is is, and really shouldn't _have_ to care).

Reply:
Yeah, basically it is:

1. `import config from 'my-app/config/environment` only works in host apps
2. `getOwner(this).resolveRegistration('config:environment')` only works with access to the container
3. Both of these are IMHO less explicit, also I always need to lookup the syntax for 2. as that is the only place I ever need it

Reply:
We already have the `ember-get-config` addon.

Reply:
Yes. However, 

a) it feels weird to require an addon to solve this.
b) there is current `Ember.testing` which is much more convenient. However, it is not encouraged to use it anymore, without having a clear replacement.
c) Having a clear import for this is also less error prone & clearer in intent (e.g. consider `config.environment === 'test'` vs. `config.environment === 'testing'` - oops). It is easier to find and can be auto-completed by an editor.

Reply:
> We already have the ember-get-config addon.

Sure, and that might be fine when you actually _need_ the configuration, but determining when you should or should not include some test support code should be a built in concept. 

Additionally, it is pretty important for these code paths to be stripped when `TESTING` is **not** true (e.g. `ember build --environment="production"`), that is not possible with `ember-get-config`.